src/coreclr/vm/prestub.cpp

```diff
@@ -361,6 +361,19 @@ PCODE MethodDesc::PrepareILBasedCode(PrepareCodeConfig* pConfig)
     STANDARD_VM_CONTRACT;
     PCODE pCode = NULL;
 
+#if defined(FEATURE_TIERED_COMPILATION)
+    bool shouldCountCalls = pConfig->GetMethodDesc()->IsEligibleForTieredCompilation();
+#if !defined(TARGET_X86)
+    if (shouldCountCalls
+        && (pConfig->GetCallerGCMode() == CallerGCMode::Preemptive
+            || (pConfig->GetCallerGCMode() == CallerGCMode::Unknown
+                && HasUnmanagedCallersOnlyAttribute())))
+    {
+        shouldCountCalls = false;
```

Comment:
It would be a bit better to disable call counting and update the optimization tier here like in `GetPrecompiledCode` before, instead of relying on the `JitSwitchedToOptimized` system that was meant to be used only for changes made by the JIT.
```suggestion
        NativeCodeVersion codeVersion = pConfig->GetCodeVersion();	
        if (codeVersion.IsDefaultVersion())	
        {	
            pConfig->GetMethodDesc()->GetLoaderAllocator()->GetCallCountingManager()->DisableCallCounting(codeVersion);	
        }	
        codeVersion.SetOptimizationTier(NativeCodeVersion::OptimizationTierOptimized);
        shouldCountCalls = false;
```